site/source/_posts/2017/11/16-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,254 @@
+        return v('div', {
+            classes: [ this.theme(css.root), css.rootFixed ]
+        }, [ 'Hello world!' ])
+    }
+}
+```
+
+The `classes` property is now just an array of strings.  Also we think that the mixed in method `theme()` is a lot more clear on what you are doing instead of `classes()`.  _Fixed_ classes are just simply added to the array, instead of being supplied to a chained function.
+
+This also means that the `@dojo/test-extras/harness` has been updated to remove the memoised class functionality, making tests far easier to author, not having to worry about replicating the order of adding classes to an expected widget render.  Also, we renamed `Themeable` to `Themed` as we felt it was better semantics for our mixins.
+
+## Widget lifecycle changes
+
+We were never completely happy with the APIs we provided for end users to interact with the widget lifecycle.  We have long felt that one of the biggest challenges with Dojo 1 Dijits was a complicated lifecycle.  In general the Dojo 2 widgeting system takes care of everything and developers really shouldn't need to worry about the lifecycle and instead just worry about rendering when requested.  There are still use cases where a developer does need to know when the virtual DOM they render is added to or removed from the DOM.
+
+In Beta 4 we have introduced two lifecycle methods to `WidgetBase`.  They are called `onAttach` and `onDetach`.  They are designed to allow a widget developer to _hook_ when a widget instance is attached to the DOM or detached from the DOM.
+
+To use them, it simply requires overriding the `onAttach` or `onDetach` methods.  For example:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+
+export default class MyWidget extends WidgetBase {
+    protected onAttach() {
+        // perform actions when attached
+    }
+
+    protected onDetach() {
+        // perform actins when detached
+    }
+
+    protected render() {
+        return v('div', { }, [ 'Hello world!' ])
+    }
+}
+```
+
+Previously, we had other capabilities which provided functionality in this area, but we have deprecated those and will remove them in the future.
+
+## Deferred properties
+
+A virtual DOM node can now provide its properties in a way that are resolved in a deferred fashion, where instead of providing an object of `VirtualDomProperties`, you can provide a function which takes a boolean flag of if the virtual DOM node has been insterted into the DOM and returns an object of `VirtualDomProperties`.  The widget rendering system will then only call the function when it needs to be provided with the properties, instead of the properties having to be fully calculated at render time.
+
+For example, you can now do something like this:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+
+export default class MyWidget extends WidgetBase {
```

Comment:
perhaps use a style css transition example to show the use? (ie change an opacity from `0` to `1` in the deferred properties.

Reply:
Being lazy, could you provide me a basic snippet, so I make sure I get it right?

Reply:
yup no problemo! something like:
```typescript
render() {
  return v('div', (inserted) => {
    return {
      classes: 'fadeInTransition',
      styles: {
        opacity: inserted ? '1' : '0'
      }
    };
  }, [ 'hello world' ]);
}
```